test/cipher.go

```diff
@@ -21,25 +24,274 @@ func HashBench(b *testing.B, hash func() hash.Hash) {
-		return cipher.NewCTR(bc,iv)
+		iv := make([]byte, bc.BlockSize())
+		return cipher.NewCTR(bc, iv)
 	})
 }
+*/
+
+// Compares the bits between two arrays returning the fraction
+// of differences. If the two arrays are not of the same length
+// no comparison is made and a -1 is returned.
+func BitDiff(a, b []byte) float64 {
+	if len(a) != len(b) {
+		return -1
+	}
+
+	mask1 := byte(1)
+	mask2 := byte(2)
+	mask3 := byte(4)
+	mask4 := byte(8)
+	mask5 := byte(16)
+	mask6 := byte(32)
+	mask7 := byte(64)
+	mask8 := byte(128)
+
+	count := 0
+	for i := 0; i < len(a); i++ {
+		if (a[i] & mask1) != (b[i] & mask1) {
+			count += 1
+		}
+		if (a[i] & mask2) != (b[i] & mask2) {
+			count += 1
+		}
+		if (a[i] & mask3) != (b[i] & mask3) {
+			count += 1
+		}
+		if (a[i] & mask4) != (b[i] & mask4) {
+			count += 1
+		}
+		if (a[i] & mask5) != (b[i] & mask5) {
+			count += 1
+		}
+		if (a[i] & mask6) != (b[i] & mask6) {
+			count += 1
+		}
+		if (a[i] & mask7) != (b[i] & mask7) {
+			count += 1
+		}
+		if (a[i] & mask8) != (b[i] & mask8) {
+			count += 1
+		}
+	}
+
+	return float64(count) / float64(len(a)*8)
+}
+
+// Tests a Cipher can encrypt and decrypt
+func BCHelloWorldHelper(t *testing.T,
+	newCipher func([]byte, ...interface{}) abstract.Cipher,
+	n int, bitdiff float64) {
+	text := []byte("Hello, World")
+	cryptsize := len(text)
+
+	bc := newCipher(nil)
+	keysize := bc.KeySize()
+
+	nkeys := make([][]byte, n)
+	ncrypts := make([][]byte, n)
+
+	for i := range nkeys {
+		nkeys[i] = make([]byte, keysize)
+		rand.Read(nkeys[i])
+		bc = newCipher(nkeys[i])
+		ncrypts[i] = make([]byte, cryptsize)
+		bc.Message(ncrypts[i], text, ncrypts[i])
+
+		bc = newCipher(nkeys[i])
+		decrypted := make([]byte, len(text))
+		bc.Message(decrypted, ncrypts[i], ncrypts[i])
+		if !bytes.Equal(text, decrypted) {
+			t.Log("Encryption / Decryption failed", i)
+			t.FailNow()
+		}
+	}
+
+	for i := range ncrypts {
+		for j := i + 1; j < len(ncrypts); j++ {
+			if bytes.Equal(ncrypts[i], ncrypts[j]) {
+				t.Log("Different keys result in same encryption")
+				t.FailNow()
+			}
+
+			res := BitDiff(ncrypts[i], ncrypts[j])
+			if res < bitdiff {
+				t.Log("Encryptions not sufficiently different:", res)
+				t.FailNow()
+			}
+		}
+	}
+}
+
+// Tests a Cipher:
+// 1) Encryption / decryption work
+// 2) Encryption / decryption with different key don't work
+// 3) Changing a bit in the ciphertext or mac results in failed mac check
+// 4) Different keys produce sufficiently random output
+func AuthenticateAndEncrypt(t *testing.T,
+	newCipher func([]byte, ...interface{}) abstract.Cipher,
+	n int, bitdiff float64, text []byte) {
+	cryptsize := len(text)
+	decrypted := make([]byte, len(text))
+
+	bc := newCipher(nil)
+	keysize := bc.KeySize()
+	hashsize := bc.HashSize()
+	mac := make([]byte, hashsize)
+
+	ncrypts := make([][]byte, n)
+	nkeys := make([][]byte, n)
+	nmacs := make([][]byte, n)
+
+	// Encrypt / decrypt / mac test
+	for i := range nkeys {
+		nkeys[i] = make([]byte, keysize)
```

Comment:
After more careful examination, I realized I may have been misunderstanding you. You want the key generation to be done once in a separate init style function and store the results to global variables. I can easily implement that, but then we'd have a separate helper function that performs basically the following:

```
var keys [][]byte
func TestCipherInit(key_count int, key_size int) {
  keys = make([][]byte, key_count)
  for i := 0; i < key_count; i++ {
    keys[i] = make([]byte, key_size)
    rand.Read(keys[i])
  }
}
```

We'd still need to zero out the ncrypts and nmacs in each test case, since values are Xored with existing content and in not assigned. Perhaps I'm missing something, but I think this type of change would make the test code harder to read and it doesn't seem to be very critical to performance.